Google.Api.Gax.Grpc/GrpcChannelOptions.cs

```diff
@@ -1,45 +1,282 @@
 ﻿/*
- * Copyright 2019 Google LLC
+ * Copyright 2020 Google LLC
  * Use of this source code is governed by a BSD-style
  * license that can be found in the LICENSE file or at
  * https://developers.google.com/open-source/licenses/bsd
  */
 
-using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
 
 namespace Google.Api.Gax.Grpc
 {
+    // TODO: How do you clear one of these properties?
+
     /// <summary>
-    /// Convenience methods for obtaining channel options.
+    /// Portable abstraction of channel options
     /// </summary>
-    internal static class GrpcChannelOptions
+    public sealed class GrpcChannelOptions : IEquatable<GrpcChannelOptions>
     {
+        private static readonly IReadOnlyList<CustomOption> s_emptyOptions = new List<CustomOption>().AsReadOnly();
+
+        /// <summary>
+        /// An empty set of channel options.
+        /// </summary>
+        public static GrpcChannelOptions Empty { get; } = new GrpcChannelOptions();
+
+        private GrpcChannelOptions(
+            bool? enableServiceConfigResolution = null,
+            TimeSpan? keepAlive = null,
+            string primaryUserAgent = null,
+            int? maxSendMessageSize = null,
+            int? maxReceiveMessageSize = null,
+            IReadOnlyList<CustomOption> customOptions = null)
+        {
+            EnableServiceConfigResolution = enableServiceConfigResolution;
+            KeepAlive = keepAlive;
+            PrimaryUserAgent = primaryUserAgent;
+            MaxSendMessageSize = maxSendMessageSize;
+            MaxReceiveMessageSize = maxReceiveMessageSize;
+            CustomOptions = customOptions ?? s_emptyOptions;
+        }
+
+        /// <summary>
+        /// If non-null, explicitly enables or disables service configuration resolution.
+        /// </summary>
+        public bool? EnableServiceConfigResolution { get; }
+
+        /// <summary>
+        /// If non-null, explicitly specifies the keep-alive period for the channel.
+        /// </summary>
+        public TimeSpan? KeepAlive { get; }
```

Comment:
nit: note that grpc C core supports mutliple keepalive-related options:
https://github.com/grpc/grpc/blob/5791495f595b30545f3dfd31321f4b5760b11c75/include/grpc/impl/codegen/grpc_types.h#L227
so let's make sure that the name "KeepAlive" isn't too general in case some of the other options are needed later.

Reply:
Good call. Will change to KeepAliveTime for now, to match the C-core name. We'll probably want to revisit all the naming though.